src/rpcserver.cpp

```diff
@@ -0,0 +1,196 @@
+/******************************************************************************\
+ * Copyright (c) 2021
+ *
+ * Author(s):
+ *  dtinth
+ *
+ ******************************************************************************
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation; either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
+ *
+ \******************************************************************************/
+
+#include "rpcserver.h"
+
+CRpcServer::CRpcServer ( QObject* parent, int iPort ) : QObject ( parent ), iPort ( iPort ), pTransportServer ( new QTcpServer ( this ) )
+{
+    connect ( pTransportServer, &QTcpServer::newConnection, this, &CRpcServer::OnNewConnection );
+}
+
+CRpcServer::~CRpcServer()
+{
+    if ( pTransportServer->isListening() )
+    {
+        qInfo() << "- stopping RPC server";
+        pTransportServer->close();
+    }
+}
+
+void CRpcServer::Start()
+{
+    if ( iPort < 0 )
+    {
+        return;
+    }
+    if ( pTransportServer->listen ( QHostAddress ( "127.0.0.1" ), iPort ) )
+    {
+        qInfo() << "- JSON RPC server started on port" << pTransportServer->serverPort();
+    }
+    else
+    {
+        qInfo() << "- unable to start JSON RPC server:" << pTransportServer->errorString();
+    }
+}
+
+QJsonObject CRpcServer::CreateJsonRpcError ( int code, QString message )
+{
+    QJsonObject error;
+    error["code"]    = QJsonValue ( code );
+    error["message"] = QJsonValue ( message );
+    return error;
+}
+
+QJsonObject CRpcServer::CreateJsonRpcErrorReply ( int code, QString message )
+{
+    QJsonObject object;
+    object["jsonrpc"] = QJsonValue ( "2.0" );
+    object["error"]   = CreateJsonRpcError ( code, message );
+    return object;
+}
+
+void CRpcServer::OnNewConnection()
+{
+    QTcpSocket* pSocket = pTransportServer->nextPendingConnection();
+    if ( !pSocket )
+    {
+        return;
+    }
+
+    qInfo() << "- JSON RPC accepted connection from:" << pSocket->peerAddress().toString();
+    vecClients.append ( pSocket );
+
+    connect ( pSocket, &QTcpSocket::disconnected, [this, pSocket]() {
+        qInfo() << "- JSON RPC connection from:" << pSocket->peerAddress().toString() << "closed";
+        vecClients.removeAll ( pSocket );
```

Comment:
```suggestion
        vecClients.removeAll ( pSocket );
        pSocket->deleteLater();
```

`pSocket->deleteLater();` should be added to prevent memory leaks.

Reply:
Confirmed memory leak…

- Before fix, after 10000 calls, memory usage goes up to 200+ MB

  ![image](https://user-images.githubusercontent.com/193136/132845567-a1a65b5b-00ef-4f48-b163-4c12d4bf00f0.png)

- After fix, memory usage stays the same

  ![image](https://user-images.githubusercontent.com/193136/132845612-ae696418-c495-4a84-a172-3c6ba6a19ba3.png)